From f9d318118332749c3fc5e95f85c12c6b9cb98d8d Mon Sep 17 00:00:00 2001
From: Praveen Talari <quic_ptalari@quicinc.com>
Date: Wed, 17 Jul 2024 19:28:53 +0530
Subject: [PATCH] PENDING: spi: geni-qcom: Enable SPI on SA8255p Qualcomm
 platforms

Enable or disable resources such as clocks, interconnect
paths, and TLMM(GPIOs) via the logical power domain,
invoked as part of geni_spi_transition_d3d0 where d0 is
on and d3 is off states.

The frequency is managed through the logical performance
domain with the help of performance levels, configured
by geni_spi_set_level.

As part of geni_spi_set_level, the nearest or equal operating
performance point (OPP) for the requested frequency is
obtained using dev_pm_opp_find_freq_floor() and the performance
level is set using dev_pm_opp_set_opp().

Change-Id: Icc00e341d5d1ca98f56e06fbcb234f9094c97256
Signed-off-by: Praveen Talari <quic_ptalari@quicinc.com>
Signed-off-by: Shazad Hussain <quic_shazhuss@quicinc.com>
---
 drivers/spi/spi-geni-qcom.c | 243 ++++++++++++++++++++++++++++--------
 1 file changed, 194 insertions(+), 49 deletions(-)

diff --git a/drivers/spi/spi-geni-qcom.c b/drivers/spi/spi-geni-qcom.c
index 8a217028b552..54920800e4c6 100644
--- a/drivers/spi/spi-geni-qcom.c
+++ b/drivers/spi/spi-geni-qcom.c
@@ -93,6 +93,12 @@ void spi_trace_log(struct device *dev, const char *fmt, ...)
 	va_end(args);
 }
 
+struct geni_spi_desc {
+	struct dev_pm_domain_attach_data pd_data;
+	int (*geni_se_set_rate)(struct device *dev, unsigned long clk_freq);
+	int (*geni_se_switch_state)(struct device *dev, bool state);
+};
+
 struct spi_geni_master {
 	struct geni_se se;
 	struct device *dev;
@@ -119,6 +125,7 @@ struct spi_geni_master {
 	struct dma_chan *tx;
 	struct dma_chan *rx;
 	int cur_xfer_mode;
+	const struct geni_spi_desc *dev_data;
 };
 
 static void spi_slv_setup(struct spi_geni_master *mas)
@@ -375,11 +382,19 @@ static void spi_setup_word_len(struct spi_geni_master *mas, u16 mode,
 		      __func__, bits_per_word, pack_words, word_len);
 }
 
-static int geni_spi_set_clock_and_bw(struct spi_geni_master *mas,
-					unsigned long clk_hz)
+/*
+ * geni_serial_set_rate - set clock frequency value
+ * @dev: pointer to device structure.
+ * @clk_hz: clock frequency value.
+ *
+ * return: 0 on success otherwise error.
+ */
+static int geni_spi_set_rate(struct device *dev, unsigned long clk_hz)
 {
-	u32 clk_sel, m_clk_cfg, idx, div;
+	struct spi_controller *spi = dev_get_drvdata(dev);
+	struct spi_geni_master *mas = spi_controller_get_devdata(spi);
 	struct geni_se *se = &mas->se;
+	u32 clk_sel, m_clk_cfg, idx, div;
 	int ret;
 
 	if (clk_hz == mas->cur_speed_hz)
@@ -407,11 +422,18 @@ static int geni_spi_set_clock_and_bw(struct spi_geni_master *mas,
 
 	/* Set BW quota for CPU as driver supports FIFO mode only. */
 	se->icc_paths[CPU_TO_GENI].avg_bw = Bps_to_icc(mas->cur_speed_hz);
-	ret = geni_icc_set_bw(se);
-	if (ret)
-		return ret;
+	return geni_icc_set_bw(se);
+}
 
-	return 0;
+static int geni_spi_set_clock_and_bw(struct spi_geni_master *mas,
+				     unsigned long clk_hz)
+{
+	int ret = 0;
+
+	if (mas->dev_data->geni_se_set_rate)
+		ret = mas->dev_data->geni_se_set_rate(mas->se.dev, clk_hz);
+
+	return ret;
 }
 
 static int setup_fifo_params(struct spi_device *spi_slv,
@@ -1074,13 +1096,98 @@ static irqreturn_t geni_spi_isr(int irq, void *data)
 	return IRQ_HANDLED;
 }
 
+/*
+ * geni_serial_set_level - set perf level
+ * @dev: pointer to device structure.
+ * @clk_freq: clock frequency value.
+ *
+ * return: 0 on success otherwise error.
+ */
+static int geni_spi_set_level(struct device *dev, unsigned long clk_freq)
+{
+	struct spi_controller *spi = dev_get_drvdata(dev);
+	struct spi_geni_master *mas = spi_controller_get_devdata(spi);
+	struct device *perf_dev = mas->se.pd_list->pd_devs[DOMAIN_IDX_PERF];
+	struct dev_pm_opp *opp;
+	int ret;
+
+	if (!perf_dev)
+		return -ENODEV;
+
+	/*
+	 * Find the nearest frequency level for the requested frequency.
+	 */
+	opp = dev_pm_opp_find_freq_floor(perf_dev, &clk_freq);
+	if (IS_ERR(opp)) {
+		dev_err(dev, "failed to find opp for freq %lu\n", clk_freq);
+		return PTR_ERR(opp);
+	}
+
+	ret = dev_pm_opp_set_opp(perf_dev, opp);
+	dev_pm_opp_put(opp);
+	return ret;
+}
+
+/*
+ * geni_spi_transition_d3d0 - transition from d0 -> d3 or d3 ->d0
+ * @dev: pointer to device structure.
+ * @state: decides whether it is D0 -> D3 or D3 ->D0.
+ *	true	: d3 --> d0 (on)
+ *	false	: d0 --> d3 (off)
+ * return: 0 on success otherwise error.
+ */
+static int geni_spi_transition_d3d0(struct device *dev, bool state)
+{
+	struct spi_controller *spi = dev_get_drvdata(dev);
+	struct spi_geni_master *mas = spi_controller_get_devdata(spi);
+
+	return geni_se_transition_d3d0(&mas->se, state);
+}
+
+/*
+ * geni_spi_switch_resource_state - switch resources from on -> off or off -> on
+ * @dev: pointer to dev structure.
+ * @state: decides whether it is on or off.
+ *
+ * return: 0 on success otherwise error.
+ */
+static int geni_spi_switch_resource_state(struct device *dev, bool state)
+{
+	struct spi_controller *spi = dev_get_drvdata(dev);
+	struct spi_geni_master *mas = spi_controller_get_devdata(spi);
+	int ret;
+
+	if (state) {
+		ret = geni_icc_enable(&mas->se);
+		if (ret)
+			return ret;
+
+		ret = geni_se_resources_on(&mas->se);
+		if (ret) {
+			geni_icc_disable(&mas->se);
+			return ret;
+		}
+
+		ret = dev_pm_opp_set_rate(mas->dev, mas->cur_sclk_hz);
+	} else {
+		/* Drop the performance state vote */
+		dev_pm_opp_set_rate(dev, 0);
+
+		ret = geni_se_resources_off(&mas->se);
+		if (ret)
+			return ret;
+
+		ret = geni_icc_disable(&mas->se);
+	}
+	return ret;
+}
+
 static int spi_geni_probe(struct platform_device *pdev)
 {
 	int ret, irq;
 	struct spi_controller *spi;
 	struct spi_geni_master *mas;
 	void __iomem *base;
-	struct clk *clk;
 	struct device *dev = &pdev->dev;
 
 	irq = platform_get_irq(pdev, 0);
@@ -1095,10 +1202,6 @@ static int spi_geni_probe(struct platform_device *pdev)
 	if (IS_ERR(base))
 		return PTR_ERR(base);
 
-	clk = devm_clk_get(dev, "se");
-	if (IS_ERR(clk))
-		return PTR_ERR(clk);
-
 	spi = devm_spi_alloc_host(dev, sizeof(*mas));
 	if (!spi)
 		return -ENOMEM;
@@ -1110,16 +1213,46 @@ static int spi_geni_probe(struct platform_device *pdev)
 	mas->se.dev = dev;
 	mas->se.wrapper = dev_get_drvdata(dev->parent);
 	mas->se.base = base;
-	mas->se.clk = clk;
 
-	ret = devm_pm_opp_set_clkname(&pdev->dev, "se");
-	if (ret)
-		return ret;
-	/* OPP table is optional */
-	ret = devm_pm_opp_of_add_table(&pdev->dev);
-	if (ret && ret != -ENODEV) {
-		dev_err(&pdev->dev, "invalid OPP table in device tree\n");
-		return ret;
+	mas->dev_data = device_get_match_data(&pdev->dev);
+	if (!mas->dev_data)
+		return -ENXIO;
+
+	if (of_device_is_compatible(dev->of_node, "qcom,sa8255p-geni-spi")) {
+		ret = dev_pm_domain_attach_list(&pdev->dev, &mas->dev_data->pd_data,
+						&mas->se.pd_list);
+		if (ret < 0)
+			return ret;
+
+		ret = pm_runtime_resume_and_get(mas->se.pd_list->pd_devs[DOMAIN_IDX_PERF]);
+		if (ret)
+			goto spi_geni_pm_domain_detach;
+	} else {
+		mas->se.clk = devm_clk_get(mas->se.dev, "se");
+		if (IS_ERR(mas->se.clk))
+			return PTR_ERR(mas->se.clk);
+
+		ret = devm_pm_opp_set_clkname(mas->se.dev, "se");
+		if (ret)
+			return ret;
+
+		/* OPP table is optional */
+		ret = devm_pm_opp_of_add_table(mas->se.dev);
+		if (ret && ret != -ENODEV) {
+			dev_err(mas->se.dev, "invalid OPP table in device tree, ret:%d\n", ret);
+			return ret;
+		}
+
+		ret = geni_icc_get(&mas->se, NULL);
+		if (ret)
+			return ret;
+		/* Set the bus quota to a reasonable value for register access */
+		mas->se.icc_paths[GENI_TO_CORE].avg_bw = Bps_to_icc(CORE_2X_50_MHZ);
+		mas->se.icc_paths[CPU_TO_GENI].avg_bw = GENI_DEFAULT_BW;
+
+		ret = geni_icc_set_bw(&mas->se);
+		if (ret)
+			return ret;
 	}
 
 	spi->bus_num = -1;
@@ -1152,22 +1285,14 @@ static int spi_geni_probe(struct platform_device *pdev)
 	pm_runtime_set_autosuspend_delay(&pdev->dev, 250);
 	ret = devm_pm_runtime_enable(dev);
 	if (ret)
-		return ret;
+		goto spi_geni_pm_domain_detach;
 
 	if (device_property_read_bool(&pdev->dev, "spi-slave"))
 		spi->target = true;
 
-	/* Set the bus quota to a reasonable value for register access */
-	mas->se.icc_paths[GENI_TO_CORE].avg_bw = Bps_to_icc(CORE_2X_50_MHZ);
-	mas->se.icc_paths[CPU_TO_GENI].avg_bw = GENI_DEFAULT_BW;
-
-	ret = geni_icc_set_bw(&mas->se);
-	if (ret)
-		return ret;
-
 	ret = spi_geni_init(mas);
 	if (ret)
-		return ret;
+		goto spi_geni_pm_domain_detach;
 
 	/*
 	 * check the mode supported and set_cs for fifo mode only
@@ -1196,6 +1321,8 @@ static int spi_geni_probe(struct platform_device *pdev)
 	free_irq(mas->irq, spi);
 spi_geni_release_dma:
 	spi_geni_release_dma_chan(mas);
+spi_geni_pm_domain_detach:
+	dev_pm_domain_detach_list(mas->se.pd_list);
 	return ret;
 }
 
@@ -1210,41 +1337,43 @@ static void spi_geni_remove(struct platform_device *pdev)
 	free_irq(mas->irq, spi);
 
 	spi_geni_release_dma_chan(mas);
+	dev_pm_domain_detach_list(mas->se.pd_list);
 }
 
 static int __maybe_unused spi_geni_runtime_suspend(struct device *dev)
 {
 	struct spi_controller *spi = dev_get_drvdata(dev);
 	struct spi_geni_master *mas = spi_controller_get_devdata(spi);
-	int ret;
+	int ret = 0;
 
 	trace_spi_info(mas->dev, __func__, "start");
-	/* Drop the performance state vote */
-	dev_pm_opp_set_rate(dev, 0);
-
-	ret = geni_se_resources_off(&mas->se);
-	if (ret)
-		return ret;
+	if (mas->dev_data->geni_se_switch_state) {
+		ret = mas->dev_data->geni_se_switch_state(dev, false);
+		if (ret) {
+			dev_err(dev, "failed to switch resources, ret:%d\n", ret);
+			return ret;
+		}
+	}
 
-	return geni_icc_disable(&mas->se);
+	return ret;
 }
 
 static int __maybe_unused spi_geni_runtime_resume(struct device *dev)
 {
 	struct spi_controller *spi = dev_get_drvdata(dev);
 	struct spi_geni_master *mas = spi_controller_get_devdata(spi);
-	int ret;
+	int ret = 0;
 
 	trace_spi_info(mas->dev, __func__, "start");
-	ret = geni_icc_enable(&mas->se);
-	if (ret)
-		return ret;
-
-	ret = geni_se_resources_on(&mas->se);
-	if (ret)
-		return ret;
+	if (mas->dev_data->geni_se_switch_state) {
+		ret = mas->dev_data->geni_se_switch_state(dev, true);
+		if (ret) {
+			dev_err(dev, "failed to switch resources, ret:%d\n", ret);
+			return ret;
+		}
+	}
 
-	return dev_pm_opp_set_rate(mas->dev, mas->cur_sclk_hz);
+	return ret;
 }
 
 static int __maybe_unused spi_geni_suspend(struct device *dev)
@@ -1287,8 +1416,24 @@ static const struct dev_pm_ops spi_geni_pm_ops = {
 	SET_SYSTEM_SLEEP_PM_OPS(spi_geni_suspend, spi_geni_resume)
 };
 
+static const struct geni_spi_desc geni_spi = {
+	.geni_se_set_rate = geni_spi_set_rate,
+	.geni_se_switch_state = geni_spi_switch_resource_state,
+};
+
+static const struct geni_spi_desc remotely_geni_spi = {
+	.pd_data = {
+		.pd_flags = PD_FLAG_NO_DEV_LINK,
+		.pd_names = (const char*[]) { "power", "perf" },
+		.num_pd_names = 2,
+	},
+	.geni_se_set_rate = geni_spi_set_level,
+	.geni_se_switch_state = geni_spi_transition_d3d0,
+};
+
 static const struct of_device_id spi_geni_dt_match[] = {
-	{ .compatible = "qcom,geni-spi" },
+	{ .compatible = "qcom,geni-spi", .data = &geni_spi },
+	{ .compatible = "qcom,sa8255p-geni-spi", .data = &remotely_geni_spi },
 	{}
 };
 MODULE_DEVICE_TABLE(of, spi_geni_dt_match);
-- 
2.17.1

